### ISSUE TITLE
feat(workflows): dynamic DAG — spawn new tasks based on runtime measurement results

### ISSUE THREAD

**Issue by ddri** [marqov-dev/marqov-sdk#30]
## The Gap

The current `@workflow` decorator builds a static DAG at definition time. All tasks and their dependencies must be known before any task runs. This means the orchestrator cannot spawn new tasks based on intermediate quantum measurement results.

This is a hard requirement for adaptive variational algorithms, now the recommended approach for quantum chemistry.

## Why This Matters

**ADAPT-VQE** (Grimsley et al., *Nature Communications* 2019, 500+ citations) is the canonical example. Its outer loop:

1. Measure energy gradients for every operator in a pool — O(100–1000) parallel circuit evaluations
2. Classically select the operator with the largest gradient norm
3. Append that operator to the ansatz circuit
4. Re-run VQE inner loop with the new circuit
5. Check convergence — if not converged, repeat from step 1

Steps 2–5 require the orchestrator to construct a new task (with a structurally different circuit) based on the results of step 1. The number of outer iterations is unknown at submission time. This cannot be expressed as a static DAG.

The same pattern appears in: ADAPT-QAOA, quantum reinforcement learning, adaptive state preparation, and quantum machine learning training loops.

## What the Literature Says

**Qurator** (arXiv:2604.05505, April 2026) models hybrid workloads as dynamic DAGs with explicit quantum semantics. Its key finding: no existing tool jointly optimises queue time and circuit fidelity across providers. The dynamic DAG framing is how Qurator represents workloads internally, enabling it to make scheduling decisions that provider-locked tools cannot.

**Pilot-Quantum** (arXiv:2604.03445 and arXiv:2412.18519) proposes a four-layer middleware architecture and introduces "execution motifs" — recurring interaction patterns (tightly-coupled variational loops, loosely-coupled task-parallel circuit batches, circuit-cutting pipelines) that dictate different scheduling requirements. The key differentiator it identifies between frameworks is **runtime resource elasticity** — the ability to add or reallocate QPU/CPU resources mid-execution as queue conditions change.

**Tierkreis** (arXiv:2211.02350, Quantinuum, 2022) was designed specifically to fix the static-DAG problem using a higher-order dataflow graph where subgraphs can be passed as values, enabling runtime DAG construction.

## The Minimum Viable Primitive

The entire class of adaptive algorithms reduces to one operation:

> Given the result of task N, conditionally submit task N+1 with a modified circuit or updated parameters.

Everything more complex (ADAPT-style circuit growth, RL-guided workflows) builds on this primitive.

## Important Distinction: Three Layers of "Dynamic"

This issue is about **layer 1 only**:

1. **Classical workflow branching** ← this issue — run circuit, get results, decide what to run next (seconds timescale, orchestrator layer)
2. **Mid-circuit feed-forward** — within a single circuit execution, use a measurement result to conditionally apply a gate (microseconds, FPGA layer). IBM Heron, Quantinuum H-series, IonQ Forte support this. The SDK should pass these constructs through transpilation without stripping them — separate issue.
3. **QEC feedback loops** — nanosecond cycles managed by dedicated hardware. Not an SDK concern.

## Design Considerations

**Determinism.** Any imperative dynamic workflow body must be deterministic and replay-safe — no wall-clock reads, no unseeded RNG outside tasks. This is a constraint Temporal imposes and any implementation must surface clearly.

**Reproducibility.** Dynamic workflows make the executed DAG data-dependent. The orchestrator should record the realised execution trace — every spawned task, its inputs, and the measurement results that drove each branch — so a dynamic run is fully auditable and replayable against recorded results. This turns a gap-filling feature into a positioning statement: dynamic + auditable is something Tierkreis and Covalent don't make first-class.

**Cost boundedness.** A static DAG has a knowable cost envelope at submission; a dynamic workflow with unknown iteration count does not. Budget envelopes (max iterations plus per-iteration cost caps) are required so this feature composes with preflight checks and spend limits rather than bypassing them.

**Queue affinity.** Each dynamically spawned circuit re-enters provider queues, so on shared hardware ADAPT-VQE wall-time is queue-dominated. Provider session mechanisms (Qiskit Runtime Sessions, Braket Hybrid Jobs) mitigate this and are an adjacent requirement — possibly a separate issue.

## Proposed API Direction (for discussion)

Two candidate designs — community input welcome:

**Option A: Imperative dynamic workflow**
```python
@dynamic_workflow  # replay-safe; no side effects outside tasks
async def adapt_vqe(initial_circuit, operator_pool, max_iterations=20):
    circuit = initial_circuit
    for _ in range(max_iterations):
        gradients = await fan_out_gradient_measurements(circuit, operator_pool)
        if converged(gradients):
            break
        circuit = circuit.append(select_operator(gradients))
    return circuit
```

**Option B: Continuation-passing**
Tasks return a `WorkflowDispatch` that the engine executes next, building the DAG one step at a time.

The exact API is open for design — this issue is to establish the requirement and gather community input.

## References

- Grimsley et al., "An adaptive variational algorithm for exact molecular simulations," *Nature Communications* 10, 3007 (2019)
- Tang et al., "qubit-ADAPT-VQE," *PRX Quantum* (2021) — arXiv:1911.10205
- "Recent Developments in VQE: Survey and Benchmarking" (arXiv:2602.11384, 2026)
- Qurator: arXiv:2604.05505
- Pilot-Quantum: arXiv:2604.03445 and arXiv:2412.18519
- Tierkreis: arXiv:2211.02350
- "Harnessing Patterns to Support the Development of Hybrid Quantum Applications" (arXiv:2507.00696)
- "Mining Quantum Software Patterns in Open-Source Projects" (arXiv:2601.06281)

**Comment by Zaneham** [comment]
Hey David! Had a proper dig through the workflows package, really clean setup. Quick thought on #30 that might make it lighter than it looks.

The thing blocking dynamic DAGs isn't really Temporal, it's the eager graph capture sitting in front of it. Because `@workflow` runs the body once with task proxies to freeze the graph, the task outputs are placeholders at build time, so there's nowhere to branch on a real measurement result. `JobWorkflow` then just walks the precomputed `execution_levels`. Totally sensible for static graphs, but that's where the dynamism gets lost.

The nice part is you're already sitting on the thing that solves it. Temporal natively does durable, replayable, imperative workflows, await an activity, look at the result, conditionally schedule the next one, and it's all replay-safe (you're already leaning on `workflow.now() ` for exactly that reason). So for the adaptive case like ADAPT-VQE I'd let those be first-class Temporal workflows rather than pre-baked graphs. The body stays imperative, which is your Option A, and it's safe not because you built a journal but because Temporal's event history *is* the journal. That also hands you the "record the realized execution trace" requirement for free, since the event history is the trace.

So it lands on A, but specifically A-as-a-real-Temporal-workflow, going with the grain of the tool instead of fighting it. The honest tradeoff is you lose `visualize()` on the dynamic ones, you can't draw a graph that doesn't exist yet, so you'd reconstruct the realized DAG from the event history afterward. I'd keep eager-capture `@workflow` for the static cases where the upfront picture matters and graduate the adaptive ones to native Temporal.

(For what it's worth, this whole shape, durable conditional execution with a replayable trace, is the old checkpoint/restart pattern from the mainframe world, so there's decades of prior art all pointing the same way: lean on the engine's log, don't roll your own.)

Always happy to prototype the adaptive path on a small ADAPT-VQE if it's useful, squarely the kind of thing I like poking at.

Edit: punctuation/minor tidbits
